test: compile all unit tests into a single executable

syuu1228 opened the pull request:
To reduce test executable size and speed up compilation time, compile unit
tests into a single executable.

Closes https://github.com/scylladb/scylladb/issues/9155

avikivity commented:
We should wait with this until we switch to cmake, to avoid having to repeat the work on cmake.

As I noted before, this is anti-modular. To test anything (say test/boost/expr_test.cc) we have to build everything.

syuu1228 commented:
> We should wait with this until we switch to cmake, to avoid having to repeat the work on cmake.

Okay, switching to CMake build system is this PR https://github.com/scylladb/scylladb/pull/20188, right?

@tchaikov How can I test my patch with cmake build system, just passing --use-cmake to configure.py is enough? 
Also, you mentioned "Note: The native Ninja generation in configure.py will be removed after a grace period to ensure smooth transition.", how can I check this affects my patch or not?

tchaikov commented:
> > We should wait with this until we switch to cmake, to avoid having to repeat the work on cmake.
> 
> Okay, switching to CMake build system is this PR #20188, right?

yes.

> 
> @tchaikov How can I test my patch with cmake build system, just passing --use-cmake to configure.py is enough?

i think so. in short, passing `--use-cmake` instructs `configure.py` to generate `build.ninja` using CMake.

> Also, you mentioned "Note: The native Ninja generation in configure.py will be removed after a grace period to ensure smooth transition.", how can I check this affects my patch or not?

this is but a warning which informs developers whose workflow(s) is wired to, for instance, the build directories or the paths used by the existing `build.ninja` rules generated by `configure.py` directly. but after the "grace period", i will just drop the code in `configure.py` which is not shared by the CMake-based building system. if you want to verify your change with the CMake-based building system. as you noticed, passing `--use-cmake` would do the trick. and `test.py` has been adapted to use the artifacts built with the CMake-based building system.

syuu1228 commented:
Rebased with latest master since some of tests are conflicted.
Also, implemented single executable support for CMake mode.
Since dependencies information defined on configure.py is not referenced on CMake mode, we need to define them directly on test/boost/CMakeLists.txt:
https://github.com/scylladb/scylladb/compare/48ffada09207a4e0344f58a2e2e94d6e58e5bbf6..e4beac219c46aae5bcd6a6fda13bcafef5d68326

tchaikov commented on test/boost/CMakeLists.txt:
```diff
@@ -1,384 +1,200 @@
-  KIND SEASTAR)
-add_scylla_test(schema_loader_test
+add_scylla_test(boost_combined_test
+  SOURCES boost_combined_test.cc
```
both `SOURCES` and `LIBRARIES` are so called multi-value arguments. see https://github.com/scylladb/scylladb/blob/3fda9651ccbcb7b23f843515297683bca091253b/test/CMakeLists.txt#L11 . and we use [cmake_parse_arguments](https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html) to parse the arguments. in general, one can just use something like
```cmake
add_scylla_test(boost_combined_test
  SOURCES
    boost_combined_test.cc
    UUID_test.cc
    # more sources here
  LIBRARIES
    auth
    alternator
    # more libraries here
    Seastar::seastar_testing
  KIND BOOST)
```
to define a test.

in which, i'd suggest 

- add indent to the arguments passed to the keywords like `SOURCES`
- order the source files alphabetically
- group the external libraries and internal/convenience libraries separately

tchaikov commented:
> Rebased with latest master since some of tests are conflicted. Also, implemented single executable support for CMake mode. Since dependencies information defined on configure.py is not referenced on CMake mode, we need to define them directly on test/boost/CMakeLists.txt: https://github.com/scylladb/scylladb/compare/48ffada09207a4e0344f58a2e2e94d6e58e5bbf6..e4beac219c46aae5bcd6a6fda13bcafef5d68326

i don't follow this part. do you mean you are adding more linked libraries in order to build the test? i am asking, because these tests did build last week using CMake.

avikivity commented:
As I mentioned previously, this is anti-modular. We won't be able to test a single module without rebuilding the entire system.

denesb commented:
> As I mentioned previously, this is anti-modular. We won't be able to test a single module without rebuilding the entire system.

So should we stop all intermediate improvements to our test building and work towards modules (which might be years away) to solve problems we face today?
I understand that you don't want the codebase to move in the opposite direction, but as I tried to explain [above](https://github.com/scylladb/scylladb/pull/21443#issuecomment-2467660580), I don't think that is the case here, despite how the title may suggest that.

avikivity commented:
> > As I mentioned previously, this is anti-modular. We won't be able to test a single module without rebuilding the entire system.
> 
> So should we stop all intermediate improvements to our test building and work towards modules (which might be years away) to solve problems we face today? 

I don't agree that parallelizing test.py on many machines is a good direction. Our CI is already a monster and it will become even less scrutable.

> I understand that you don't want the codebase to move in the opposite direction, but as I tried to explain [above](https://github.com/scylladb/scylladb/pull/21443#issuecomment-2467660580), I don't think that is the case here, despite how the title may suggest that.

Then compile only tests that depend on the entire system (cql_test_env) into a single executable.

denesb commented:
> Then compile only tests that depend on the entire system (cql_test_env) into a single executable.

Ok, I'm fine with that as a middle ground, looks like 1/3 of our tests use `cql_test_env`.
```
$ cd test/boost
$ ls -l |*_test.cc  wc -l
179
$ grep *_test.cc -l -e cql_test_env | wc -l
60
```

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13749/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13749/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/UUID_test
:small_blue_diamond: boost/address_map_test
:small_blue_diamond: boost/aggregate_fcts_test
:small_blue_diamond: boost/allocation_strategy_test
:small_blue_diamond: alternator
:small_blue_diamond: boost/anchorless_list_test
:small_blue_diamond: boost/auth_passwords_test
:small_blue_diamond: boost/auth_resource_test
:small_blue_diamond: boost/auth_test
:small_blue_diamond: boost/aws_error_injection_test
:small_blue_diamond: boost/aws_errors_test
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: boost/big_decimal_test
:small_blue_diamond: boost/bloom_filter_test
:small_blue_diamond: boost/bptree_test
:small_blue_diamond: boost/broken_sstable_test
:small_blue_diamond: boost/btree_test
:small_blue_diamond: boost/bytes_ostream_test
:small_blue_diamond: boost/cache_algorithm_test
:small_blue_diamond: boost/cache_mutation_reader_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/620/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13318/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2194/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5311/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13749/)
#### Build Details:
- Duration: 6 hr 42 min
- Builder: spider4.cloudius-systems.com

syuu1228 commented:
> > Then compile only tests that depend on the entire system (cql_test_env) into a single executable.
> 
> Ok, I'm fine with that as a middle ground, looks like 1/3 of our tests use `cql_test_env`.
> 
> ```
> $ cd test/boost
> $ ls -l |*_test.cc  wc -l
> 179
> $ grep *_test.cc -l -e cql_test_env | wc -l
> 60
> ```

@denesb So the next step is dropping the test which does not depends on cpl_test_env from the single executable and revert them to the previous normal executable?

denesb commented:
> > > Then compile only tests that depend on the entire system (cql_test_env) into a single executable.
> > 
> > 
> > Ok, I'm fine with that as a middle ground, looks like 1/3 of our tests use `cql_test_env`.
> > ```
> > $ cd test/boost
> > $ ls -l |*_test.cc  wc -l
> > 179
> > $ grep *_test.cc -l -e cql_test_env | wc -l
> > 60
> > ```
> 
> @denesb So the next step is dropping the test which does not depends on cpl_test_env from the single executable and revert them to the previous normal executable?

Yes, looks like this is the compromise we can make to move this forward.

syuu1228 commented:
> > > > Then compile only tests that depend on the entire system (cql_test_env) into a single executable.
> > > 
> > > 
> > > Ok, I'm fine with that as a middle ground, looks like 1/3 of our tests use `cql_test_env`.
> > > ```
> > > $ cd test/boost
> > > $ ls -l |*_test.cc  wc -l
> > > 179
> > > $ grep *_test.cc -l -e cql_test_env | wc -l
> > > 60
> > > ```
> > 
> > 
> > @denesb So the next step is dropping the test which does not depends on cpl_test_env from the single executable and revert them to the previous normal executable?
> 
> Yes, looks like this is the compromise we can make to move this forward.

Implemented this.

avikivity commented:
Please adjust CMakeLists.txt so it works the same way and test with --use-cmake.

Also, the patch changelog should report before/after binary sizes.

denesb commented:
I wonder how does invoking a certain test from this monolith binary will look like?
Will we be able to target all tests from a certain file (test-case)? I think it should work, because previously boost created an implicit per-file test-case for use and now we instantiate a test-case manually per file.

syuu1228 commented:
Pushed new version with following changes:
- Added before/after binary size on the changelog, as @avikivity mentioned.
- Stop moving variables/functions before BOOST_AUTO_TEST_SUITE to avoid the compilation error, since it increases patch size.
  Solved the problem by closing & reopening the namespace which BOOST_AUTO_TEST_SUITE defined.
- rebased with master.

syuu1228 commented:
I did some mistake on configure.py / CMakeLists.txt, there were missing files after dropping the test which does not depends on cpl_test_env from the single executable, fixed.

syuu1228 commented:
> Please adjust CMakeLists.txt so it works the same way and test with --use-cmake.

I tested both without --use-cmake and with --use-cmake.

syuu1228 commented:
> I wonder how does invoking a certain test from this monolith binary will look like? Will we be able to target all tests from a certain file (test-case)? I think it should work, because previously boost created an implicit per-file test-case for use and now we instantiate a test-case manually per file.

It still can call tests per file.
Each unit tests are now become a "suite" on Boost Test, so we need to specify `--run_test=<suite>`.
Or if we want running only one test in a file, we can add testname after the suite name: `--run_test=<suite>/<testname>`.

Here are the example.
- Without the single executable
```
# run all tests of virtual_table_test
$ ./build/release/test/boost/virtual_table_test
# run specific test of virutal_table_test
$ ./build/release/test/boost/virtual_table_test --run_test=test_set_cell
```
- With the single executable
```
# run all tests of virtual_table_test
$ ./build/release/test/boost/combined_tests --run_test=virtual_table_test
# run specific test of virutal_table_test
$ ./build/release/test/boost/combined_tests --run_test=virtual_table_test/test_set_cell
```

Since we added each test file as a "suite",  `--list_content`  will shows up all of the tests in two levels, "suite" and "case".

- Without the single executable
```
$ ./build/release/test/boost/virtual_table_test --list_content
test_set_cell*
test_system_config_table_read*
test_system_config_table_update*
test_system_config_table_no_live_update*
```
- With the single executable
```
$ ./build/release/test/boost/combined_tests --list_content 
aggregate_fcts_test*
    test_aggregate_avg*
    test_aggregate_sum*
    test_aggregate_max*
    test_aggregate_min*
    test_aggregate_count*
    test_reverse_type_aggregation*
    test_minmax_on_set*
auth_test*
    test_default_authenticator*
    test_password_authenticator_attributes*
    test_password_authenticator_operations*
    roles_table_is_protected*
    role_members_table_is_protected*
    role_permissions_table_is_protected*
    test_alter_with_timeouts*
    test_alter_with_workload_type*
    test_try_to_create_role_with_hashed_password_and_password*
    test_try_to_create_role_with_password_and_hashed_password*
    test_try_create_role_with_hashed_password_as_anonymous_user*
    test_try_login_after_creating_roles_with_hashed_password*
    test_try_describe_schema_with_internals_and_passwords_as_anonymous_user*

...

virtual_reader_test*
    test_query_size_estimates_virtual_table*
    test_query_view_built_progress_virtual_table*
virtual_table_test*
    test_set_cell*
    test_system_config_table_read*
    test_system_config_table_update*
    test_system_config_table_no_live_update*
```

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14019/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14019/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/aggregate_fcts_test
:small_blue_diamond: boost/auth_test
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: boost/cache_algorithm_test
:small_blue_diamond: boost/castas_fcts_test
:small_blue_diamond: boost/cdc_test
:small_blue_diamond: boost/column_mapping_test
:small_blue_diamond: boost/commitlog_cleanup_test
:small_blue_diamond: boost/commitlog_test
:small_blue_diamond: boost/cql_auth_query_test
:small_blue_diamond: boost/cql_functions_test
:small_blue_diamond: boost/cql_query_group_test
:small_blue_diamond: boost/cql_query_large_test
:small_blue_diamond: boost/cql_query_like_test
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/data_listeners_test
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/error_injection_test
:small_blue_diamond: boost/extensions_test
:small_blue_diamond: boost/filtering_test
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5573/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/239/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13586/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2456/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/837/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14019/)
#### Build Details:
- Duration: 5 hr 12 min
- Builder: spider6.cloudius-systems.com

avikivity commented on configure.py:
```diff
@@ -430,181 +430,122 @@ def find_ninja():
 }
 
 scylla_tests = set([
```
> It reduces executable sizes 5.5GB on release, and 10GB on debug.

nice

avikivity commented:
The suites even show in test.py log!

asias commented on test.py:
```diff
@@ -362,18 +368,56 @@ class BoostTestSuite(UnitTestSuite):
+        # Apply combined test only for test/boost
+        if self.name != 'boost':
+            return
+        exe = path_to(self.mode, "test", self.name, 'combined_tests')
```
This makes the other boost tests to fail if combined_tests is not built.

$ ninja build/debug/test/boost/big_decimal_test
$ ./test.py -s --repeat 1 --mode debug -v  boost/big_decimal_test

```
[5/6] LINK (stripped) build/debug/test/boost/big_decimal_test
Traceback (most recent call last):
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 2252, in workaround_python26789
    code = await main()
           ^^^^^^^^^^^^
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 1946, in main
    await find_tests(options)
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 1597, in find_tests
    suite = TestSuite.opt_create(f, options, mode)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 217, in opt_create
    suite = SpecificTestSuite(path, cfg, options, mode)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 401, in __init__
    self._generate_cache()
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 377, in _generate_cache
    exe = path_to(self.mode, "test", self.name, 'combined_tests')
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/asias/src/cloudius-systems/scylla/./test.py", line 100, in path_to
    raise FileNotFoundError(f"{exe_path} does not exist.")
FileNotFoundError: build/debug/test/boost/combined_tests does not exist.
```

asias replied:
@syuu1228 FYI.
